frontends/pytorch/csrc/builder/ivalue_importer.cpp

```diff
@@ -19,22 +19,78 @@
+// Python terms, what we want here is a hash of `id(x)` -- unfortunately, IValue
+// is not uniformly heap allocated the way a `PyObject*` is, so special handling
+// is needed. At the time of this writing, there seem to be two different
+// implementations, neither of which is exactly what we want.
+//
+// - c10::IValue::hash static method
+//   - Problem: Doesn't handle certain data types, in particular objects (which
+//   modules are a special case of) and lists/dicts. This makes sense when
+//   reflecting the Python semantics.
+// - c10::WeakIValue::hash method
+//   - Problem: it literally just returns the bits of the "union" as an int.
+//   This seems to read uninitialized bits for the bool variant.
+//
+// We use the `c10::IValue::hash` static method with special cases for data
+// types that need their identity to be handled specially. `c10::IValue::hash`
+// seems to be implemented in a principled way following the Python semantics,
+// which is compatible with the semantics we want (for the subset it doesn't
+// throw an error on).
+namespace {
+struct IValueHasher {
+  size_t operator()(const c10::IValue &ivalue) const {
+    if (ivalue.isObject() || ivalue.isList()) {
+      return std::hash<const void *>()(
+          static_cast<const void *>(ivalue.internalToPointer()));
+    }
+
+    return c10::IValue::hash(ivalue);
+  }
+};
+} // namespace
+
+// TODO: The implementation of isSameIdentity looks vulnerable to malloc reusing
+// the same memory block (if this hash function is used in an online setting,
+// such as when tracing). Can we do better?
+namespace {
+struct IValueEq {
+  bool operator()(const c10::IValue &lhs, const c10::IValue &rhs) const {
+    return lhs.isSameIdentity(rhs);
+  }
+};
+} // namespace
+
 namespace {
 /// Helper class for holding state during recursive IValue import.
 ///
 /// The intended usage pattern of this class is to construct it then call
-/// `importIValue` exactly once. Calling `importIValue` more than once
-/// is likely to produce unexpected results since the same in-memory IValue
-/// can be reimported more than once. That is, subsequent calls to
-/// `importIValue` will not properly unify IValue's with already-imported
-/// IValue's.
+/// `importIValue`.
 ///
-/// TODO: Support unifying repeated IValue's.
-/// This already is an issue when importing a single IValue through the current
-/// API, because the same underlying Tensor (or List/Dict) can be referenced by
-/// multiple properties of a module. There is an extra complication with tensors
-/// because they can alias each other in fairly arbitrary ways, which we will
-/// need to model with slice ops.
+/// The `importIValue` method can be called more than once, and values are
+/// unified *by object identity*. For types isomorphic to Python builtin types
+/// the behavior is what you would expect from `id(x)`.
+///
+/// For tensors, object identity is a little tricky. As background, at::Tensor
```

Comment:
Thanks for this explanation. I feel like the ACAP side is suffering from a bug due to mis-applying some of this...